orttraining/orttraining/core/graph/pipeline_transformer.cc

```diff
@@ -963,229 +1091,417 @@ common::Status SplitGraph(Graph& graph,
+  auto op_assigned = std::find(std::begin(stages), std::end(stages), -1);
+  ORT_RETURN_IF_NOT(op_assigned == std::end(stages), 
+                    "All ops must be assigned to a stage");
+
+  // Edges always go forward.
+  for (size_t i = 0, t = graph.MaxNodeIndex(); i < t; ++i) {
+    Node* node = graph.GetNode(i);
+    int node_stage = stages[i];
+    auto& node_outputs = node->MutableOutputDefs();
+    for (NodeArg* arg : node_outputs) {
+      if (arg == nullptr || !arg->HasTensorOrScalarShape())
+        continue;
+      auto cs = graph.GetMutableConsumerNodes(arg->Name());
+      for (Node* c : cs) {
+        int outgoing_stage = stages[c->Index()];
+        ORT_RETURN_IF_NOT(node_stage <= outgoing_stage);
+      }
+    }
+  }
 
-  // Split the graph by cutting edges specified in cut_info. After this function, the graph will be
-  // composed of several disconnected partitions.
-  ORT_RETURN_IF_ERROR(SplitGraph(graph, cut_info, send_nodes, recv_nodes));
+  return Status::OK();
+}
 
-  if (send_nodes.size() != split_count || recv_nodes.size() != split_count) {
-    return ORT_MAKE_STATUS(ONNXRUNTIME, FAIL, "Split error: not all cut has Send and Recv inserted. Send node count: ",
-                           send_nodes.size(), ", Recv node count: ", recv_nodes.size(), ", split count: ", split_count);
+// TODO: optimize and verify.
+Status GetDeviceAssignmentMap(Graph& graph, 
+                              const std::map<std::string, int>& id_to_stage,
+                              std::map<Node*, int>& op_to_stage) {
+  for (size_t i = 0, t = graph.MaxNodeIndex(); i < t; ++i) {
+    Node* node = graph.GetNode(i);
+    bool found = false;
+    auto& node_outputs = node->MutableOutputDefs();
+    for (NodeArg* arg : node_outputs) {
+      if (id_to_stage.find(arg->Name()) != id_to_stage.end()) {
+        int stage = id_to_stage.at(arg->Name());
+        op_to_stage.insert({node, stage});
+        found = true;
+        break;
+      }
+    }
+    ORT_ENFORCE(found, "Can't find node's stage " + node->Name());
   }
 
-  // Check to see if there are any initializers that is being shared between different partitions. If there
-  // is, keep the initializer in the first seen partition and have it pass through by send/recv to the others.
-  ORT_RETURN_IF_ERROR(HandleSharedInitializer(graph, send_nodes, recv_nodes));
+  // TODO: call verify assignment.
+  return Status::OK();
+}
 
-  // Now remove the partitions that are not tie to the current pipeline stage and generate the sub-graph.
-  if (pipeline_stage_id < split_count) {
-    ORT_RETURN_IF_ERROR(GenerateSubgraph(graph, send_nodes[pipeline_stage_id]));
-  } else {
-    ORT_RETURN_IF_ERROR(GenerateSubgraph(graph, recv_nodes.back()));
+
+Status GetDeviceAssignmentMap(Graph& graph,
+                              const std::vector<TrainingSession::TrainingConfiguration::CutInfo>& cuts,
+                              std::map<Node*, int>& op_to_stage) {
+  auto total_nodes = graph.MaxNodeIndex();
```

Comment:
What about using Graph::NumberOfNodes here ?
https://github.com/microsoft/onnxruntime/blob/master/include/onnxruntime/core/graph/graph.h#L727
See the comment about 'removed' nodes in the header.